src/backend/sgx/enclave.rs

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Enclave object
+
+use std::fs::{metadata, File, OpenOptions};
+use std::io::{Error, ErrorKind, Result};
+use std::os::unix::fs::FileTypeExt;
+use std::path::Path;
+
+pub const SGX_ENCLAVE_DEVICE: &str = "/dev/sgx_enclave";
+
+/// Wraps an enclave file descriptor.
+pub struct Enclave(File);
+
+impl From<Enclave> for File {
+    fn from(enclave: Enclave) -> File {
+        enclave.0
+    }
+}
+
+impl From<File> for Enclave {
+    fn from(file: File) -> Enclave {
+        Self(file)
+    }
+}
+
+impl Enclave {
+    pub fn new() -> Result<Enclave> {
+        if !Path::new(SGX_ENCLAVE_DEVICE).exists() {
+            return Err(Error::new(
+                ErrorKind::NotFound,
+                format!("{}: not found", SGX_ENCLAVE_DEVICE),
+            ));
+        }
+
+        if !metadata(SGX_ENCLAVE_DEVICE)
+            .unwrap()
+            .file_type()
+            .is_char_device()
+        {
+            return Err(Error::new(ErrorKind::NotFound, "{}: not a chracter device"));
+        }
+
+        OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open(SGX_ENCLAVE_DEVICE)
+            .map(Into::into)
```

Comment:
This function has two tests followed by an open. The first test is not needed. Instead, we should check for the error code that indicates that the open failed because the file was missing.

I don't understand the purpose of the second test. What are we defending against here? The ioctls will fail if it isn't the right device. So what is the concern?

Reply:
not hugely important